### ISSUE TITLE
windows install with pre-built libraries

### ISSUE THREAD

**Issue by Pfedj** [dzop/emacs-zmq#11]
windows 10
emacs 27.0.50
zmq-20190408.1552

I have compiled emacs with `--with-modules`  option . Then I download the pre-built libraries from last release (29 Mar) and put them in root zmq project. 
I have added `(require 'zmq)` in my init file.
Eval with `emacs --debug-init`:

```
Debugger entered--Lisp error: (module-open-failed "c:/Users/PC/.emacs.d/elpa/zmq-20190408.1552/emacs-zmq.dll" "\315\345 \355\340\351\344\345\355 \363\352\340\347\340\355\355\373\351 \354\356\344\363\353\374.")
  load("c:/Users/PC/.emacs.d/elpa/zmq-20190408.1552/emacs-zmq.dll" nil nil t)
  load-file("c:/Users/PC/.emacs.d/elpa/zmq-20190408.1552/emacs-zmq.dll")
  zmq-load()
  byte-code("\300 \210\301\302!\207" [zmq-load provide zmq] 2)
  require(zmq)
  eval-buffer(#<buffer  *load*> nil "c:/Users/PC/.emacs.d/init.el" nil t)  ; Reading at buffer position 6107
  load-with-code-conversion("c:/Users/PC/.emacs.d/init.el" "c:/Users/PC/.emacs.d/init.el" t t)
  load("c:/Users/PC/.emacs.d/init" noerror nomessage)
  startup--load-user-init-file(#f(compiled-function () #<bytecode 0x115d5df>) #f(compiled-function () #<bytecode 0x115d5af>) t)
  command-line()
  normal-top-level()
```
Could you help me with this issue?

**Comment by nnicandro** [comment]
Can you try building with Emacs 26.1 to see if its an issue with your Emacs 27 build. The Emacs available from `pacman` appears to work since the tests pass https://ci.appveyor.com/project/dzop/emacs-zmq. Can you try installing Emacs using the MSYS2 system using the command 

```
pacman --noconfirm -S mingw-w64-x86_64-emacs
```

as mentioned in the test script https://github.com/dzop/emacs-zmq/blob/appveyor/appveyor.yml.

Also if you are downloading this package for use with `emacs-jupyter`, there will be other problems you will encounter even if you get this package working that I am still in the process of figuring out. See #10.

**Comment by Pfedj** [comment]
I have installed Emacs 26.1 with:
```
  pacman --noconfirm -S mingw-w64-x86_64-emacs
```
Also I have took `dll` with: 
```
wget https://github.com/dzop/emacs-zmq/releases/download/v0.10.1/emacs-zmq-x86_64-w64-mingw32.tar.gz

tar -xzf emacs-zmq-x86_64-w64-mingw32.tar.gz
```

And all tests was passed successfully:
```
make EMACS=/mingw64/bin/emacs test
```
==> `Ran 11 tests, 11 results as expected`
But with start Emacs, nothing has changed here:
```
  load("c:/Users/PC/emacs-zmq/emacs-zmq.dll" nil nil t)
  load-file("c:/Users/PC/emacs-zmq/emacs-zmq.dll")
  (progn (load-file zmq-module-file) (add-hook 'post-gc-hook (function zmq--cleanup-globrefs)))
  (if (file-exists-p zmq-module-file) (progn (load-file zmq-module-file) (add-hook 'post-gc-hook (function zmq--cleanup-globrefs))) (if (y-or-n-p "ZMQ module not found. Build it? ") (progn (let ((default-directory (file-name-directory (locate-library "zmq"))) (emacs (if (and invocation-directory invocation-name) (progn (file-truename (expand-file-name invocation-name invocation-directory)))))) (let (--cl-load-zmq--) (setq --cl-load-zmq-- (function (lambda (_buf status) (progn (if (string= status "finished\n") (zmq-load) (message "Something went wrong when compiling the ZMQ module!")) (remove-hook 'compilation-finish-functions --cl-load-zmq--) (exit-recursive-edit))))) (add-hook 'compilation-finish-functions --cl-load-zmq--) (compile (concat "make" (if emacs (progn (concat " EMACS=" emacs))))) (recursive-edit))))))
  (if zmq-module-file (if (file-exists-p zmq-module-file) (progn (load-file zmq-module-file) (add-hook 'post-gc-hook (function zmq--cleanup-globrefs))) (if (y-or-n-p "ZMQ module not found. Build it? ") (progn (let ((default-directory (file-name-directory (locate-library "zmq"))) (emacs (if (and invocation-directory invocation-name) (progn (file-truename (expand-file-name invocation-name invocation-directory)))))) (let (--cl-load-zmq--) (setq --cl-load-zmq-- (function (lambda (_buf status) (progn (if (string= status "finished\n") (zmq-load) (message "Something went wrong when compiling the ZMQ module!")) (remove-hook 'compilation-finish-functions --cl-load-zmq--) (exit-recursive-edit))))) (add-hook 'compilation-finish-functions --cl-load-zmq--) (compile (concat "make" (if emacs (progn (concat " EMACS=" emacs))))) (recursive-edit)))))) (user-error "Modules are not supported"))
  zmq-load()
  eval-buffer(#<buffer  *load*-786045> nil "c:/Users/PC/emacs-zmq/zmq.el" nil t)  ; Reading at buffer position 22257
  load-with-code-conversion("c:/Users/PC/emacs-zmq/zmq.el" "c:/Users/PC/emacs-zmq/zmq.el" nil nil)
  load("c:/Users/PC/emacs-zmq/zmq.el" nil nil t)
  load-file("c:/Users/PC/emacs-zmq/zmq.el")
  load-user-file("~/emacs-zmq/zmq.el")
  eval-buffer(#<buffer  *load*> nil "c:/Users/PC/.emacs.d/init.el" nil t)  ; Reading at buffer position 6042
  load-with-code-conversion("c:/Users/PC/.emacs.d/init.el" "c:/Users/PC/.emacs.d/init.el" t t)
  load("c:/Users/PC/.emacs.d/init" t t)
  #f(compiled-function () #<bytecode 0x1000c0c55>)()
  command-line()
  normal-top-level()
```
My init config now:
```
(add-to-list 'load-path "~/emacs-zmq/") 
(load-user-file "~/emacs-zmq/zmq.el")   
(require 'zmq)
```

**Comment by nnicandro** [comment]
Are you using `/mingw64/bin/emacs` when you get the error? If the tests pass, then Emacs was able to load the module otherwise the tests would not have succeeded. So it looks like you must be using a different Emacs than `/mingw64/bin/emacs` when you get that error.

**Comment by Pfedj** [comment]
I have checked. It is the same Emac. It doesn't work with your dll.
On my other computer  I have compiled emacs  27.050 with --with-modules option. I tried this version with your dll. In this case everything works properly ( both tests and execution of emacs).
Also i noticed different conclusion when doing tests. In case of emacs 27.050 (that works) when doing test i have the following different outcome:
```
$ make EMACS=/mingw64/bin/emacs test
/mingw64/bin/emacs -nw -Q -batch -L . -l ert -l zmq-tests.el \
--eval "(ert-run-tests-batch-and-exit)"
Loading c:/Users/Pfedj/emacs-zmq/emacs-zmq.dll (module)...
Running 11 tests (2019-05-01 12:29:46+0300, selector `t')
   passed   1/11  zmq-close-socket-once (0.014011 sec)
   passed   2/11  zmq-contexts (0.008442 sec)
   passed   3/11  zmq-encryption (0.000171 sec)
   passed   4/11  zmq-globrefs (0.490145 sec)
   passed   5/11  zmq-messages (0.032373 sec)
   passed   6/11  zmq-polling (2.066917 sec)
   passed   7/11  zmq-send-unicode (0.058721 sec)
   passed   8/11  zmq-sockets (0.035505 sec)
   passed   9/11  zmq-subprocess (0.569780 sec)
   passed  10/11  zmq-utility (0.018173 sec)
   passed  11/11  zmq-wrong-object-type (0.003131 sec)

Ran 11 tests, 11 results as expected (2019-05-01 12:29:49+0300, 3.299473 sec)

```
In case of 26.1 when doing test the following outcome:
```
$ make EMACS=/mingw64/bin/emacs test
/mingw64/bin/emacs -nw -Q -batch -L . -l ert -l zmq-tests.el \
--eval "(ert-run-tests-batch-and-exit)"
Running 11 tests (2019-05-01 12:11:24+0300)
   passed   1/11  zmq-close-socket-once
   passed   2/11  zmq-contexts
   passed   3/11  zmq-encryption
   passed   4/11  zmq-globrefs
   passed   5/11  zmq-messages
   passed   6/11  zmq-polling
   passed   7/11  zmq-send-unicode
   passed   8/11  zmq-sockets
   passed   9/11  zmq-subprocess
   passed  10/11  zmq-utility
   passed  11/11  zmq-wrong-object-type

Ran 11 tests, 11 results as expected (2019-05-01 12:11:27+0300)
```

**Comment by nnicandro** [comment]
OK, I think the issue here has to do with how loading `dll` files works on Windows. When the `emacs-zmq.dll` is loaded by Emacs it looks for the `libzmq.dll` in the order specified by https://docs.microsoft.com/en-us/windows/desktop/dlls/dynamic-link-library-search-order#standard-search-order-for-desktop-applications. In the case of `make test`, it finds `libzmq.dll` in the current directory (see (5) in the link). But when you run Emacs normally it isn't able to find the `libzmq.dll` since starting Emacs normally doesn't start it in the zmq directory.

I think you have two options:
1. Copy the `libzmq.dll` to the directory containing the Emacs executable. This way (1) in the dll search order will find it.
2. In a batch file, set the `%PATH%` environment variable to include the emacs-zmq directory, before launching Emacs. (6) in the dll search order. For example

```
set %PATH%=path\to\emacs-zmq;%PATH%
runemacs.exe
```

Maybe @fleimgruber can weigh in on if he has every experienced the issue with finding the libzmq.dll when starting Emacs outside the emacs-zmq directory. When testing `emacs-jupyter` in Appveyor I found that I had to do something similar to (2) for the tests to work so I'm assuming this is an issue for all Window's users.

**Comment by fleimgruber** [comment]
@dzop AFAIK your analysis is correct. Something like this is indeed needed for all Windows users and I should have put it in the README while working on the Windows compat. When I was at it I also added the emacs-zmq directory to the internal Emacs PATH variable (exec-path) and I *think* that this is enough / a third option along the ones already mentioned. The way this is done depends on local init.el (e.g. manually, exec-path-from-shell, Spacemacs .spacemacs.env, ...).

**Comment by nnicandro** [comment]
Setting `exec-path` or using something like `setenv` only affects subprocesses that are started by Emacs. What we need to do is set the `PATH` in Emacs's parent environment since that's what the Windows function that loads the `dll` will use, this is only possible by launching Emacs in some `batch` file that pre-sets the `PATH` since it isn't possible (nor should it be) to modify the environment of a parent process.

**Comment by nnicandro** [comment]
The fix here is to link `libzmq` statically into the `emacs-zmq.dll` file so that there is no dependency on `libzmq.dll` at all. Obvious now that I that I think about it.

**Comment by fleimgruber** [comment]
Ah good to know and quite reasonable, come to think of it. Static linking would be convenient, yes.

**Comment by nnicandro** [comment]
@fleimgruber @Pfedj Can someone download the dll from the latest release on the releases page to try it out. I've statically linked `libzmq` in the `emacs-zmq.dll` binary. I've tested it on Appveyor and it works, but I would like confirmation from someone else on Windows as a sanity check.

**Comment by Pfedj** [comment]
I have cloned the last version  and get last version `emacs-zmq.dll.` `(require 'zmq)  `causes no mistakes anymore . 
However nothing happens with the latest `emacs-juputer ` when I eval  `jupyter-run-repl `.
There is the message:
```

Starting python3 kernel process...done
Kernel did not send startup message
Kernel starting up...done
Requesting kernel info...done
if: Kernel did not respond to kernel-info reques
```
It may not relate to this library and I should replace it to `jupyter-emacs.`

**Comment by nnicandro** [comment]
Yes it would be better to make a new issue over at `emacs-jupyter`. When you do, can you show the output after you run `M-: (setq jupyter--debug t) RET M-x jupyter-run-repl RET` in the new issue.

**Comment by fleimgruber** [comment]
I can confirm that the dll from the latest release with static `libzmq` works over here.

**Comment by nnicandro** [comment]
Closed in 1bc8268021e718e1e25b16b318d1e0c10fc053b7.
